File: src/Atlasd/Battlenet/BinaryReader.cs
using Atlasd.Battlenet.Exceptions;
using System.IO;
using System.Text;

namespace Atlasd.Battlenet
{
    class BinaryReader : System.IO.BinaryReader
    {
        private readonly object _lock = new object();

        public BinaryReader(Stream input) : base(input, Encoding.UTF8) { }
        public BinaryReader(Stream input, Encoding encoding) : base(input, encoding) { }
        public BinaryReader(Stream input, Encoding encoding, bool leaveOpen) : base(input, encoding, leaveOpen) { }

        public long GetNextNull()
        {
            lock (_lock)
            {
                long lastPosition = BaseStream.Position;

                while (BaseStream.Position < BaseStream.Length)
                {
                    if (ReadByte() == 0)
                    {
                        long r = BaseStream.Position;
                        BaseStream.Position = lastPosition;
                        return r;
                    }
                }

                BaseStream.Position = lastPosition;
                return -1;
            }
        }

        public byte[] ReadByteString()
        {
            lock (_lock)
            {
                var nullPos = GetNextNull();
                if (nullPos < 0)
                {
                    throw new GameProtocolViolationException(null,
                        $"Truncated string field at stream position {BaseStream.Position}: missing null terminator");
                }
                var size = nullPos - BaseStream.Position;
                return ReadBytes((int)size)[..^1];
            }
        }

        public override string ReadString()
        {
            lock (_lock)
            {
                string str = "";
                char chr;
                while (BaseStream.Position < BaseStream.Length && (int)(chr = ReadChar()) != 0)
                    str += chr;
                return str;
            }
        }
    }
}

File: src/AtlasV/Battlenet/BNET/BinaryReader.vb
Imports System.IO
Imports System.Text

Namespace AtlasV.Battlenet
    Class BinaryReader
        Inherits System.IO.BinaryReader

        Public Sub New(ByVal input As Stream)
            MyBase.New(input, Encoding.UTF8)
        End Sub

        Public Sub New(ByVal input As Stream, ByVal encoding As Encoding)
            MyBase.New(input, encoding)
        End Sub

        Public Sub New(ByVal input As Stream, ByVal encoding As Encoding, ByVal leaveOpen As Boolean)
            MyBase.New(input, encoding, leaveOpen)
        End Sub

        Public Function GetNextNull() As Long
            Dim lastPosition As Long = BaseStream.Position

            While BaseStream.CanRead

                If ReadByte() = 0 Then
                    Dim r As Long = BaseStream.Position
                    BaseStream.Position = lastPosition
                    Return r
                End If
            End While

            Return -1
        End Function

        Public Function ReadByteString() As Byte()
            Dim size = GetNextNull() - BaseStream.Position
            Return ReadBytes(size - 1)
        End Function

        Public Overrides Function ReadString() As String
            Dim str As String = ""
            Dim chr As Char = ReadChar()

            While chr <> vbNullChar
                str += chr
                chr = ReadChar()
            End While

            Return str
        End Function

    End Class

End Namespace